### PR TITLE
Refactor/ndalgebra

### PR DESCRIPTION
**Pull request by altavir**
Refactoring of NDAlgebra for #196 and a general cleanup.

### CODE REVIEW

File: examples/src/benchmarks/kotlin/kscience/kmath/benchmarks/ViktorLogBenchmark.kt [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+
+    @Benchmark
+    fun rawViktorLog() {
+        val fortyTwo = F64Array.full(dim, dim, init = 42.0)
+        var res: F64Array
+        repeat(n) {
+            res = fortyTwo.log()
+        }
+    }
+}
```
Comment (CommanderTvis) [inline examples/src/benchmarks/kotlin/kscience/kmath/benchmarks/ViktorLogBenchmark.kt:40]: Missing newline.

File: examples/src/main/kotlin/kscience/kmath/operations/ComplexDemo.kt [hunk 2]
```diff
@@ -1,18 +1,17 @@
...
     // 2d element
-    val element = NDElement.complex(2, 2) { (i,j) ->
+    val element = NDAlgebra.complex(2, 2).produce { (i,j) ->
         Complex(i.toDouble() - j.toDouble(), i.toDouble() + j.toDouble())
     }
     println(element)
 
     // 1d element operation
-    val result = with(NDField.complex(8)) {
+    val result = with(NDAlgebra.complex(8)) {
```
Comment (CommanderTvis) [inline examples/src/main/kotlin/kscience/kmath/operations/ComplexDemo.kt:14]: The `with` call seems to be redundant. (NDAlgebra.complex(...)) {}

File: examples/src/main/kotlin/kscience/kmath/structures/ComplexND.kt [hunk 3]
```diff
@@ -26,8 +27,10 @@ fun main() {
 
     val complexTime = measureTimeMillis {
         complexField {
-            var res: NDBuffer<Complex> = one
-            repeat(n) { res += 1.0 }
+            var res: NDStructure<Complex> = one
```
Comment (CommanderTvis) [inline examples/src/main/kotlin/kscience/kmath/structures/ComplexND.kt:30]: I think `UNUSED` inspection in files like this one should be suppressed.

File: examples/src/main/kotlin/kscience/kmath/operations/ComplexDemo.kt (hunk 2 above)
Comment (altavir) [reply examples/src/main/kotlin/kscience/kmath/operations/ComplexDemo.kt:14]: I do not like additional braces. With seems idiomatic here.

**Review by rgrit91 — APPROVED** [review]

**Review by CommanderTvis — APPROVED** [review]

**Pull request merged**